### ISSUE TITLE
Unable to easily remove the source message field

### ISSUE THREAD

**Issue by joshuaspence** [elastic/logstash#6101]
I apologise if I am filing this issue on the wrong repository, but I don't think that this issue is unique to [logstash-filter-json](https://github.com/logstash-plugins/logstash-filter-json). Basically, what I am trying to do is parse a JSON-encoded message and then remove the JSON-encoded field. Ideally, I would be able to achieve this with the following configuration:

```
filter {
  json {
    remove_field => ["message"]
    source => "message"
  }
}
```

This works if the JSON-encoded message doesn't contain a `message` field itself, but if it does then the wrong field is removed from the event. I was able to work around this by using the following configuration instead, but this feels hacky:

```
filter {
  if [type] == "json" {
    mutate {
      add_field => {
        "[@metadata][message]" => "%{message}",
      }
      remove_field => ["message"]
    }

    json {
      source => "[@metadata][message]"
    }
  }
}
```

**Comment by jordansissel** [comment]
Howdy! I think I understand what you are asking, and I agree the behavior is weird, but I'm not clear on one part:

> the wrong field is removed from the event

What do you mean "the wrong field is removed" ?

**Comment by joshuaspence** [comment]
Right, sorry... I understand that my wording could have been ambiguous. Let me reword my issue with an example.

Let's suppose that I have the following Logstash configuration:

```
filter {
  json {
    remove_field => ["message"]
    source => "message"
  }
}
```

And now let's also suppose that I have the following events (the entire event has been JSON-encoded for convenience):

  1. `{"message" => "{\"message\": \"This is a message\", \"severity\": \"info\"}"}`
  2. `{"message" => "{\"severity\": \"error\"}"}`

The _expected_ events output from this filter are:

  1. `{"message" => "This is a message", "severity" => "info"}`
  2. `{"severity" => "error"}`

The _actual_ events output from this filter are:

  1. `{"severity" => "info"}`
  2. `{"severity" => "error"}`

**Comment by jordansissel** [comment]
Ahh I see now. This makes sense to me now; thanks for the clarification!
